src/vos/vos_ts.c

```diff
@@ -0,0 +1,376 @@
+			miss_cursor += miss_size;
+		}
+	}
+
+	*ts_tablep = ts_table;
+
+	return 0;
+}
+
+void
+vos_ts_table_free(struct vos_ts_table **ts_tablep)
+{
+	struct vos_ts_table	*ts_table = *ts_tablep;
+
+	/** entry 0 points to start of allocated space */
+	D_FREE(ts_table->tt_table[0].te_miss_idx);
+	D_FREE(ts_table);
+
+	*ts_tablep = NULL;
+}
+
+/** This probably needs more thought */
+static bool
+ts_update_on_evict(struct vos_ts_table *ts_table, struct vos_ts_entry *entry)
+{
+	struct vos_ts_entry	*parent = NULL;
+	struct vos_ts_entry	*other = NULL;
+	struct vos_ts_info	*info = entry->te_info;
+	uint32_t		*idx;
+
+	if (entry->te_record_ptr == NULL)
+		return false;
+
+	if (entry->te_parent_ptr != NULL) {
+		parent = vos_ts_lookup_idx(ts_table, entry->te_parent_ptr);
+		if (info->ti_type & 1) { /* negative entry */
+			other = parent;
+		} else if (parent != NULL) {
+			idx = &parent->te_miss_idx[entry->te_hash_idx];
+			other = vos_ts_lookup_idx(ts_table, idx);
+		}
+	}
+
+	if (other == NULL) {
+		ts_table->tt_ts_rl = MAX(ts_table->tt_ts_rl, entry->te_ts_rl);
+		ts_table->tt_ts_rh = MAX(ts_table->tt_ts_rh, entry->te_ts_rh);
+		ts_table->tt_ts_w = MAX(ts_table->tt_ts_w, entry->te_ts_w);
+		return true;
+	}
+
+	other->te_ts_rl = MAX(other->te_ts_rl, entry->te_ts_rl);
+	other->te_ts_rh = MAX(other->te_ts_rh, entry->te_ts_rh);
+	other->te_ts_w = MAX(other->te_ts_w, entry->te_ts_w);
+
+	return true;
+}
+
+#define TS_TRACE(action, entry, idx, type)				\
+	D_DEBUG(DB_TRACE, "%s %s at idx %d(%p), read.hi="DF_U64		\
+		" read.lo="DF_U64" write="DF_U64"\n", action,		\
+		type_strs[type], idx, (entry)->te_record_ptr,		\
+		(entry)->te_ts_rh, (entry)->te_ts_rl, (entry)->te_ts_w)
+
+static inline void
+evict_one(struct vos_ts_table *ts_table, struct vos_ts_entry *entry,
+	  uint32_t idx, struct vos_ts_info *info, bool removed)
+{
+	if (ts_update_on_evict(ts_table, entry)) {
+		TS_TRACE("Evicted", entry, idx, info->ti_type);
+		entry->te_record_ptr = NULL;
+	}
+
+	if (removed)
+		return;
+
+	if (info->ti_mru == idx)
+		info->ti_mru = entry->te_prev_idx;
+
+	if (info->ti_lru == idx)
+		return;
+
+	/** Remove the entry from it's current location */
+	remove_ts_entry(&ts_table->tt_table[0], entry);
+
+	/** insert the entry at the LRU */
+	insert_ts_entry(&ts_table->tt_table[0], entry, idx, info->ti_mru,
+			info->ti_lru);
+
+	info->ti_lru = idx;
+}
+
+static inline void
+evict_children(struct vos_ts_table *ts_table, struct vos_ts_info *info,
+	       struct vos_ts_entry *entry)
+{
+	struct vos_ts_entry	*child;
+	int			 i;
+	uint32_t		 idx;
+	uint32_t		 cache_num;
+
+	info = entry->te_info;
+
+	if ((info->ti_type == VOS_TS_TYPE_AKEY) || (info->ti_type & 1) != 0)
+		return;
+
+	cache_num = info->ti_cache_mask + 1;
+	info++;
+	for (i = 0; i < cache_num; i++) {
+		/* Also evict the children, if present */
+		idx = entry->te_miss_idx[i] & VOS_TS_MASK;
+		child = &ts_table->tt_table[idx];
+		if (child->te_record_ptr != &entry->te_miss_idx[i])
```

Comment:
Yes, if the entry becomes LRU and gets evicted while caching a child entry for another record, it will be set to the address of that record’s index field.